features/prog_test.go

```diff
@@ -0,0 +1,88 @@
+package features
+
+import (
+	"fmt"
+	"math"
+	"os"
+	"testing"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/internal/testutils"
+)
+
+var progTypeMinVersion = map[ebpf.ProgramType]string{
+	ebpf.SocketFilter:          "3.19",
+	ebpf.Kprobe:                "4.1",
+	ebpf.SchedCLS:              "4.1",
+	ebpf.SchedACT:              "4.1",
+	ebpf.TracePoint:            "4.7",
+	ebpf.XDP:                   "4.8",
+	ebpf.PerfEvent:             "4.9",
+	ebpf.CGroupSKB:             "4.10",
+	ebpf.CGroupSock:            "4.10",
+	ebpf.LWTIn:                 "4.10",
+	ebpf.LWTOut:                "4.10",
+	ebpf.LWTXmit:               "4.10",
+	ebpf.SockOps:               "4.13",
+	ebpf.SkSKB:                 "4.14",
+	ebpf.CGroupDevice:          "4.15",
+	ebpf.SkMsg:                 "4.17",
+	ebpf.RawTracepoint:         "4.17",
+	ebpf.CGroupSockAddr:        "4.17",
+	ebpf.LWTSeg6Local:          "4.18",
+	ebpf.LircMode2:             "4.18",
+	ebpf.SkReuseport:           "4.19",
+	ebpf.FlowDissector:         "4.20",
+	ebpf.CGroupSysctl:          "5.2",
+	ebpf.RawTracepointWritable: "5.2",
+	ebpf.CGroupSockopt:         "5.3",
+	ebpf.Tracing:               "5.5",
+	ebpf.StructOps:             "5.6",
+	ebpf.Extension:             "5.6",
+	ebpf.LSM:                   "5.7",
+	ebpf.SkLookup:              "5.9",
+}
+
+func TestHaveProgType(t *testing.T) {
+	for progType := ebpf.UnspecifiedProgram + 1; progType <= progType.Max(); progType++ {
+		// Need inner loop copy to make use of t.Parallel()
+		pt := progType
+
+		minVersion, ok := progTypeMinVersion[pt]
+		if !ok {
+			// In cases where a new prog type wasn't added to progTypeMinVersion
+			// we should make sure the test runs anyway and fails on old kernels
+			minVersion = "0.0"
+		}
+
+		feature := fmt.Sprintf("program type %s", pt.String())
+
+		t.Run(pt.String(), func(t *testing.T) {
+			t.Parallel()
+			if pt == ebpf.LircMode2 {
+				t.Skip("LircMode2 isn't supported by testing kernels")
```

Comment:
I think we could use `testutils.SkipIfNotSupported()` here, so at least the probe runs, but we discard ErrNotSupported until/if/when we rebuild CI kernels with this kconfig enabled.

Reply:
`SkipIfNotSupported()` might be handy to create an easy test for the HaveProgHelper API. We could basically run a double for over progs and helpers and discard NotSupported results. While its not to accurate it would at least give good coverage and we could avoid creating a complicated "test case matrix".

Reply:
Yup, we should only test a few known-good and known-bad prog/helper combinations, and combinatorially test all prog/helper types to catch errors that are not `nil`/`ErrNotSupported`. Though that won't alert us for cases like StructOps where probes always fail when no BTF is specified. That's probably something we need to accept.